File: LeetCodePractice/src/unionFind/SatisfiabilityOfEqualityEquations.java
package unionFind;
/*Given an array equations of strings that represent relationships between variables, each string equations[i] has length 4 and takes one of two different forms: "a==b" or "a!=b".  Here, a and b are lowercase letters (not necessarily different) that represent one-letter variable names.

Return true if and only if it is possible to assign integers to variable names so as to satisfy all the given equations.

Example 1:
Input: ["a==b","b!=a"]
Output: false
Explanation: If we assign say, a = 1 and b = 1, then the first equation is satisfied, but not the second.  There is no way to assign the variables to satisfy both equations.
Example 2:

Input: ["b==a","a==b"]
Output: true
Explanation: We could assign a = 1 and b = 1 to satisfy both equations.
Example 3:

Input: ["a==b","b==c","a==c"]
Output: true
Example 4:

Input: ["a==b","b!=c","c==a"]
Output: false
Example 5:

Input: ["c==c","b==d","x!=z"]
Output: true

analysis:
TC O(N)
*/
public class SatisfiabilityOfEqualityEquations {
	private int[] parent;
	public boolean equationsPossible(String[] equations) {
		//all lower case alphabet
        parent = new int[26];
        for(int i = 0; i < 26; i++){
        	parent[i] = i;
        }
        for(String s : equations){
        	if(s.charAt(1) == '='){
        		union(s.charAt(0) - 'a', s.charAt(3) - 'a');
        	}
        }
        for(String s : equations){
        	if(s.charAt(1) == '!'){
        		int root1 = find(s.charAt(0) - 'a');
        		int root2 = find(s.charAt(3) - 'a');
        		if(root1 == root2){
        			return false;
        		}
        	}
        }
        return true;
    }
	private int find(int x){
		int root = x;
		while(root != parent[root]){
			root = parent[root];
		}
		while(parent[x] != root){
			int fa = parent[x];
			parent[x] = root;
			x = fa;
		}
		return root;
	}
	private void union(int a, int b){
		int root_a = find(a);
		int root_b = find(b);
		if(root_a != root_b){
			parent[root_a] = root_b;
		}
	}
}

File: LeetCodePython/SatisfiabilityOfEqualityEquations.py
"""
You are given an array of strings equations that represent relationships between variables where each string equations[i] is of length 4 and takes one of two different forms: "xi==yi" or "xi!=yi".Here, xi and yi are lowercase letters (not necessarily different) that represent one-letter variable names.

Return true if it is possible to assign integers to variable names so as to satisfy all the given equations, or false otherwise.



Example 1:

Input: equations = ["a==b","b!=a"]
Output: false
Explanation: If we assign say, a = 1 and b = 1, then the first equation is satisfied, but not the second.
There is no way to assign the variables to satisfy both equations.
Example 2:

Input: equations = ["b==a","a==b"]
Output: true
Explanation: We could assign a = 1 and b = 1 to satisfy both equations.

Constraints:
1 <= equations.length <= 500
equations[i].length == 4
equations[i][0] is a lowercase letter.
equations[i][1] is either '=' or '!'.
equations[i][2] is '='.
equations[i][3] is a lowercase letter.
"""
from typing import List


class SatisfiabilityOfEqualityEquations:
    def equationsPossible(self, equations: List[str]) -> bool:
        parent = list(range(26))

        def find_root(x):
            root = x
            while root != parent[root]:
                root = parent[root]
            while parent[x] != root:
                fa = parent[x]
                parent[x] = root
                x = fa
            return root

        def union(a, b):
            root_a, root_b = find_root(a), find_root(b)
            if root_a != root_b:
                parent[root_a] = root_b

        for s in equations:
            if s[1] == '=':
                union(ord(s[0]) - ord('a'), ord(s[3]) - ord('a'))

        for s in equations:
            if s[1] == '!':
                root_a, root_b = find_root(ord(s[0]) - ord('a')), find_root(ord(s[3]) - ord('a'))
                if root_a == root_b:
                    return False
        return True
